kythe/docs/rfc/3339.md

```diff
@@ -0,0 +1,348 @@
+    the process (and may do so at any time after the grace period ends). The
+    exit code of an analyzer process is ignored. The grace period is defined by
+    the implementation, but must be at least 10 seconds.
+
+-   If and after the analyzer closes the output stream, the analyzer process
+    must exit.
+
+-   If and after the driver closes the input stream, the analyzer process must
+    exit.
+
+-   A **corrupt frame** is one in which the length tag does not have the correct
+    form. A frame is corrupt if the newline terminator is missing, the string
+    preceding the newline terminator contains non-digits or is empty, or the
+    stream ends before the specified number of bytes is delivered.
+
+    If the analyzer receives a corrupt frame from the driver, it must exit. If
+    the driver receives a corrupt frame from the analyzer, it must close the
+    input stream (in which case the analyzer therefore must exit).
+
+-   When an analyzer process terminates, the driver MUST consider any pending
+    analysis requests to have failed, and thereafter the environment MAY retry
+    analysis of the corresponding compilation records.
+
+### Messages
+
+The following message structures are used by the protocol:
+
+-   `InitRequest`
+
+    ```
+    {
+      "protocol": <protocol>
+      "outputEncoding": <output-encoding>
+    }
+    ```
+
+-   `InitReply`
+
+    ```
+    {
+      "protocol": <protocol>
+    }
+    ```
+
+-   `AnalyzrRequest`
+
+    ```
+    {
+      "types": <analysis-type-list>
+    }
+    ```
+
+-   `AnalyzeReply`
+
+    ```
+    {
+      "id": <analysis-id>,
+      "workingDir": <file-path>
+      "inputs": <file-path-list>
+      "arguments": <argument-list>
+      "environment": <environment>
+      "output": <file-path>
+      "entryContext": <string>
+      "outputKey": <file-path>
+      "details": <details>
+    }
+    ```
+
+-   `VNameRequest'
+    ```
+    {
+	"path": <file-path>
+	"signature": <string>
+    }
+    ```
+
+-   `VNameReply`
+    ```
+    {
+	"vname": <vname>
+    }
+    ```
+
+-   `LogRequest`
+
+    ```
+    {
+      "id": <analysis-id>,
+      "message": <diagnostic>
+    }
+    ```
+
+The basic types referenced above are:
+
+-   `<analysis-id>`: an integer ID assigned by the server (JSON: number).
+-   `<bytes>`: a base64-encoded byte string (JSON: string).
+-   `<diagnostic>`: a `kythe.proto.common.Diagnostic` message (JSON: object).
+-   `<digest>`: a lower-case hex-coded SHA-256 digest (JSON: string).
+-   `<file-path>`: a file path (JSON: string).
+-   `<file-path-list>`: a list of `file-path` (JSON: list).
+-   `<vname>`: a `kythe.proto.VName` message (JSON: object).
+-   `<protocol>`: a protocol version label (JSON: string).
+-   `<analysis-type-list>`: a list of an identifier for a specific type or types of analyses.
+-   `<output-encoding>`: an identifier indicating how the output will be encoded, must be one of `json` or `protobuf` (JSON: string).
```

Comment:
Would you consider it sufficient constraint that the underlying POSIX file descriptor be a pipe? Overly-constraining?

Reply:
You can stipulate that the output is append-only—no seek, no read. That would get you most of the way there. But it's also important to fix a format for the output, too.

Reply:
So my inclination here is to use "standard" delimited proto output, and streaming concatenated JSON. Because it's a file sink, and write-only, I see no need for sync points or checksums. Does anyone have a strong opinions for/against?

Reply:
Go handles unframed JSON just fine, but other languages' JSON libraries do not. I recall that TypeScript explicitly had a problem with this, and we wound up requiring each message on a line (which is easy to do since there is no internal vertical whitespace in JSON tokens, so you need only remove optional whitespace).

On that basis I'd recommend using some framing—and if you are going to use varint-prefixed records for proto, you might as well for JSON too IMO.

Reply:
This only really matters for the driver, so I think unframed is fine. But I went with newline-delimited since that seemed most tool-friendly. It's different than delimited proto, which is length-delimited, but the tool/library support is already there for delimited proto, so I'm fine with the discrepancy.

Reply:
No, it matters for the analyzer too, which will need to be able to receive responses in the control channel shared with the driver.

Reply:
(responded to the wrong comment, replaced)

I think if you're going to choose a framing you should use the same regardless of format. There's no sensible reason to multiply code paths in this case.

Reply:
I don't see the multiple code paths. Any given analyzer will emit exactly one of {json, protobuf}; nothing is shared between those two encodings.

Protobuf libraries already easily support reading and writing delimited, so that takes care of both producer (analyzer) and consumers (driver and tools). 

For JSON, I suspect there's more support for newline-delimited universally. Certainly you can take such a file and break apart the records on the command line.

Reply:
The driver will have to have multiple code paths for reading records.

Reply:
C++ has some built-in code for handling varint-prefixed records, but Go doesn't. So they're separate concerns: Reading the next record, vs. decoding it. Only one of them needs to switch, the other is a generic data-handling routine.

Reply:
[Buffer.DecodeMessage](https://github.com/golang/protobuf/blob/master/proto/decode.go#L368)

I don't imagine a driver being written in other than C++/Go/Java.

Just to be clear, I don't feel super strongly about this.

Reply:
I can't find a standard for length-encoded JSON. Just some implementations, which run the gamut of emitting the length as uint32BE, uint32LE, or even text. If you can point me to a specification, I'll adopt that.

Reply:
I don't see any reason not to use the same encoding as for protobuf messages: varint length + message. If support for command-line tools is important, the fact that you're allowing multiple different encodings means you'll want a translation tool anyway, so we can just provide that in the Kythe repo and not worry about it. That makes the encoding rule for output simpler: A stream of varint-prefixed records, containing either JSON or protobuf wire format. You can use the filename to distinguish which is which, perhaps.